xpra/platform/posix/keyboard.py

```diff
@@ -37,6 +40,55 @@ def init_vars(self):
         super().init_vars()
         self.keymap_modifiers = None
         self.keyboard_bindings = None
+        self.__dbus_helper = DBusHelper()
+        self.__input_sources = {}
+        self._dbus_gnome_shell_eval_ism(
+            ".inputSources",
+            self._store_input_sources,
+        )
+
+    def _store_input_sources(self, input_sources):
+        log("_store_input_sources(%s)", input_sources)
+        for layout_info in input_sources.values():
+            index = int(layout_info["index"])
+            layout_variant = str(layout_info["id"])
+            layout = layout_variant.split("+", 1)[0]
+            self.__input_sources[layout] = index
+
+    def _dbus_gnome_shell_eval_ism(self, cmd, callback=None):
+        ism = "imports.ui.status.keyboard.getInputSourceManager()"
+
+        def ok_cb(success, res):
+            try:
+                if not dbus_to_native(success):
+                    log("_dbus_gnome_shell_eval_ism(%s): %s", cmd, msg)
+                    return
+                if callback is not None:
+                    callback(json.loads(dbus_to_native(res)))
+            except Exception:
+                log("_dbus_gnome_shell_eval_ism(%s)", cmd, exc_info=True)
+
+        def err_cb(msg):
+            log("_dbus_gnome_shell_eval_ism(%s): %s", cmd, msg)
+
+        self.__dbus_helper.call_function(
+            "org.gnome.Shell",
+            "/org/gnome/Shell",
+            "org.gnome.Shell",
+            "Eval",
+            [native_to_dbus(ism + cmd)],
+            ok_cb,
+            err_cb,
+        )
```

Comment:
Doesn't this trigger an error on non-gnome-shell desktops? Or is it only going to trigger `err_cb`?

Reply:
In absence of the `org.gnome.Shell` target, only the `err_cb` will be invoked which logs the failure.
The `DBusHelper.call_function` stops and calls `err_cb` if

1. Target bus name/path does not exists at all (`self.bus.get_object(bus_name, path)`),
2. It exists, but they have removed the chosen function or an ACL is blocking our access to that function (`obj.get_dbus_method(function, interface)`),
3. That function is accessible, but fails to run due to any other internal reason (`fn.call_async`).